File: packages/claude-plugins/worktree/scripts/lib/env.ts
import { readFileSync, writeFileSync, existsSync, readdirSync } from 'node:fs'
import { join, basename } from 'node:path'
import type { SlotConfig } from './slots.js'

export interface EnvVar {
    key: string
    value: string
    type: 'slot' | 'copy' | 'static'
}

export function getTemplateFiles(configDir: string): string[] {
    if (!existsSync(configDir)) return []
    return readdirSync(configDir)
        .filter(f => f.startsWith('.env') && f.endsWith('.template'))
        .map(f => join(configDir, f))
}

export function getOutputFileName(templatePath: string): string {
    const name = basename(templatePath)
    return name.replace(/\.template$/, '')
}

export function parseTemplate(content: string): EnvVar[] {
    const lines = content.split('\n')
    const vars: EnvVar[] = []

    for (const line of lines) {
        const trimmed = line.trim()
        if (trimmed === '' || trimmed.startsWith('#')) continue

        const match = line.match(/^([A-Z_][A-Z0-9_]*)=(.*)$/)
        if (!match) continue

        const [, key, value] = match

        if (value.match(/\{\{[A-Z_][A-Z0-9_]*\}\}/)) {
            vars.push({ key, value, type: 'slot' })
        } else if (value.match(/\{\{COPY:[A-Z_][A-Z0-9_]*\}\}/)) {
            vars.push({ key, value, type: 'copy' })
        } else {
            vars.push({ key, value, type: 'static' })
        }
    }

    return vars
}

export function extractSlotVars(content: string): string[] {
    const matches = content.matchAll(/\{\{([A-Z_][A-Z0-9_]*)\}\}/g)
    const vars = new Set<string>()
    for (const match of matches) {
        if (!match[1].startsWith('COPY:')) {
            vars.add(match[1])
        }
    }
    return [...vars]
}

export function extractCopyVars(content: string): string[] {
    const matches = content.matchAll(/\{\{COPY:([A-Z_][A-Z0-9_]*)\}\}/g)
    const vars = new Set<string>()
    for (const match of matches) {
        vars.add(match[1])
    }
    return [...vars]
}

export function readEnvFile(path: string): Map<string, string> {
    const vars = new Map<string, string>()
    if (!existsSync(path)) return vars

    const content = readFileSync(path, 'utf-8')
    for (const line of content.split('\n')) {
        const trimmed = line.trim()
        if (trimmed === '' || trimmed.startsWith('#')) continue

        const match = line.match(/^([A-Z_][A-Z0-9_]*)=(.*)$/)
        if (match) {
            vars.set(match[1], match[2])
        }
    }
    return vars
}

export function processTemplate(
    templateContent: string,
    slotConfig: SlotConfig,
    mainEnvVars: Map<string, string>
): { content: string; warnings: string[] } {
    const warnings: string[] = []
    let result = templateContent

    // Replace slot vars {{VAR}}
    result = result.replace(/\{\{([A-Z_][A-Z0-9_]*)\}\}/g, (match, varName) => {
        if (varName.startsWith('COPY:')) return match
        const value = slotConfig[varName]
        if (value === undefined) {
            warnings.push(`Template has {{${varName}}} but not defined in slots.yaml`)
            return match
        }
        return String(value)
    })

    // Replace copy vars {{COPY:VAR}}
    result = result.replace(/\{\{COPY:([A-Z_][A-Z0-9_]*)\}\}/g, (match, varName) => {
        const value = mainEnvVars.get(varName)
        if (value === undefined) {
            warnings.push(`{{COPY:${varName}}} not found in main .env, leaving empty`)
            return ''
        }
        return value
    })

    return { content: result, warnings }
}

export function writeEnvFile(path: string, content: string): void {
    writeFileSync(path, content)
}

export function findMainEnvFiles(mainRepoDir: string): string[] {
    if (!existsSync(mainRepoDir)) return []
    return readdirSync(mainRepoDir)
        .filter(f => f.startsWith('.env') && !f.endsWith('.template') && !f.endsWith('.example'))
        .map(f => join(mainRepoDir, f))
}

File: packages/towles-tool/worktree/skills/lib/env.py
"""Environment template processing."""
from __future__ import annotations

import re
from pathlib import Path
from typing import Any, Optional


def get_template_files(config_dir: Path) -> list[Path]:
    if not config_dir.exists():
        return []
    return sorted(p for p in config_dir.iterdir() if p.name.startswith(".env") and p.name.endswith(".template"))


def get_output_filename(template_path: Path) -> str:
    return template_path.name.removesuffix(".template")


def extract_slot_vars(content: str) -> list[str]:
    """Extract {{VAR}} placeholders (not COPY:)."""
    matches = re.findall(r"\{\{([A-Z_][A-Z0-9_]*)\}\}", content)
    return list(set(v for v in matches if not v.startswith("COPY:")))


def extract_copy_vars(content: str) -> list[str]:
    """Extract {{COPY:VAR}} placeholders."""
    matches = re.findall(r"\{\{COPY:([A-Z_][A-Z0-9_]*)\}\}", content)
    return list(set(matches))


def read_env_file(path: Path) -> dict[str, str]:
    """Parse .env file into dict."""
    if not path.exists():
        return {}

    env_vars = {}
    for line in path.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        match = re.match(r"^([A-Z_][A-Z0-9_]*)=(.*)$", line)
        if match:
            env_vars[match.group(1)] = match.group(2)
    return env_vars


def process_template(
    template_content: str,
    slot_config: Optional[Any],  # SlotConfig or None
    main_env_vars: dict[str, str],
) -> tuple[str, list[str]]:
    """Process template, replacing {{VAR}} and {{COPY:VAR}}.

    If slot_config is None, slot variables are left as-is with warnings.
    Returns (processed_content, warnings).
    """
    warnings = []
    result = template_content

    # Replace slot vars {{VAR}}
    def replace_slot_var(match: re.Match) -> str:
        var_name = match.group(1)
        if var_name.startswith("COPY:"):
            return match.group(0)
        if slot_config is None:
            warnings.append(f"Template has {{{{{var_name}}}}} but no slot config available")
            return match.group(0)
        value = slot_config.get(var_name)
        if value is None:
            warnings.append(f"Template has {{{{{var_name}}}}} but not defined in slots.config.json")
            return match.group(0)
        return str(value)

    result = re.sub(r"\{\{([A-Z_][A-Z0-9_]*)\}\}", replace_slot_var, result)

    # Replace copy vars {{COPY:VAR}}
    def replace_copy_var(match: re.Match) -> str:
        var_name = match.group(1)
        value = main_env_vars.get(var_name)
        if value is None:
            warnings.append(f"{{{{COPY:{var_name}}}}} not found in main .env, leaving empty")
            return ""
        return value

    result = re.sub(r"\{\{COPY:([A-Z_][A-Z0-9_]*)\}\}", replace_copy_var, result)

    return result, warnings


def find_main_env_files(main_repo_dir: Path) -> list[Path]:
    """Find .env files in main repo (not templates or examples)."""
    if not main_repo_dir.exists():
        return []
    return sorted(
        p for p in main_repo_dir.iterdir()
        if p.name.startswith(".env")
        and not p.name.endswith(".template")
        and not p.name.endswith(".example")
        and p.is_file()
    )
